onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_builder.cc

```diff
@@ -38,37 +38,89 @@
   input_indices.push_back(operand_indices.at(input2));  // input 2
 
   if (add_activation) {
     ADD_SCALAR_OPERAND(model_builder, input_indices, fuse_code);
   }
 
-  ORT_RETURN_IF_ERROR(shaper.Eltwise(input1, input2, output));
-  const OperandType output_operand_type(operand_types.at(input1).type, shaper[output],
+  const auto output_shape = GetShapeInfoFromNodeArg(model_builder.GetGraphViewer(), output);
+  const OperandType output_operand_type(operand_types.at(input1).type, output_shape,
                                         output_scale, output_zero_point);
   ORT_RETURN_IF_ERROR(model_builder.AddOperation(op_type, input_indices,
                                                  {output}, {output_operand_type}));
   return Status::OK();
 }
 
-static Status AddSqueezeOp(ModelBuilder& model_builder,
-                           const std::string& node_name,
-                           const std::string& input, const std::string& output,
-                           std::vector<int32_t> axes) {
+static Status AddNnapiBatchNormalization(ModelBuilder& model_builder,
+                                         const std::string& input1,
+                                         const std::string& input2,
+                                         const std::string& output1,
+                                         const std::string& input3,
+                                         const std::string& output2,
+                                         int32_t fuse_code,
+                                         float output_scale = 0.0f,
+                                         int32_t output_zero_point = 0) {
+  auto& shaper(model_builder.GetShaper());
+  const auto& operand_indices(model_builder.GetOperandIndices());
+  const auto& operand_types(model_builder.GetOperandTypes());
+
+  // Add Nnapi Mul
+  InlinedVector<uint32_t> input_indices;
+  input_indices.push_back(operand_indices.at(input1));
+  input_indices.push_back(operand_indices.at(input2));
+
+  ADD_SCALAR_OPERAND(model_builder, input_indices, ANEURALNETWORKS_FUSED_NONE);
+
+  auto calculate_eltwise_output_shape = [&](const std::string& input1_name,
+                                            const std::string& input2_name,
+                                            const std::string& output_name) -> Status {
+    const Shape& shape1 = shaper[input1_name];
+    const Shape& shape2 = shaper[input2_name];
+
+    // broadcasting support
+    Shape output_shape;
+    ORT_RETURN_IF_ERROR(op_builder_helpers::PerformBroadcasting(shape1, shape2, output_shape));
+    shaper.AddShape(output_name, output_shape);
+    return Status::OK();
+  };
+
+  ORT_RETURN_IF_ERROR(calculate_eltwise_output_shape(input1, input2, output1));
+  const OperandType output_operand_type(operand_types.at(input1).type, shaper[output1],
+                                        output_scale, output_zero_point);
+  ORT_RETURN_IF_ERROR(model_builder.AddOperation(ANEURALNETWORKS_MUL, input_indices,
```

Comment:
Can you describe how/when you know to use the shaper instance to get a shape?

I see at two usage patterns but maybe there are more. 

One is here where we add something to the shaper in the lambda only to retrieve it immediately after. Is that an optional usage of Shaper and we could just provide a Shape& to the lambda or is there an implicit reason why we must call AddShape?

The other is in places like ConvOpBuilder::AddToModelBuilderImpl where we retrieve the shape of a weight from Shaper, but it's not clear to me when that was added to shaper or how I would know I'm meant to get it from there vs. the NodeArg. I assume these are cases where we've added an initializer with a different layout, but that's really opaque. 

The advantage of Shaper::operator[] calling GetShapeInfoFromNodeArg is that someone updating the NNAPI EP code can always use that to (hopefully) get a valid result. i.e. it abstracts away that a shape might come from values added with AddShape or from NodeArg shape info.

Reply:
> The advantage of Shaper::operator[] calling GetShapeInfoFromNodeArg is that someone updating the NNAPI EP code can always use that to (hopefully) get a valid result. i.e. it abstracts away that a shape might come from values added with AddShape or from NodeArg shape info.

I agree that the approach of using Shaper::operator[] to call GetShapeInfoFromNodeArg internally is much more neat and can make it a more consistent looking especially when it comes to cases like retrieving shape of weight initializer from shaper. but  I thought we think it's not a good idea to keep a dangling graph_viewer reference to Shaper, so I reverted the GetShapeInfoFromNodeArg() call in operator[] as it's not viable to call now. (maybe there's a work around for this?)

> Is that an optional usage of Shaper and we could just provide a Shape& to the lambda or is there an implicit reason why we must call AddShape?

For this particular case of BatchNorm op, Yes. don't really have to call AddShape() here. will modify it.

> Can you describe how/when you know to use the shaper instance to get a shape?

Generally speaking, where we used `shaper[]` to retrieve shape now are all special cases I think.

1.  Cases as you said where we used helper function to add initializer (to be specific weight) in different layouts. like `AddInitializerInNewLayOut()` - which further called https://github.com/microsoft/onnxruntime/blob/9edc9465f0258fa4cf83d7975905afa9ce1ec436/onnxruntime/core/providers/nnapi/nnapi_builtin/builders/model_builder.cc#L450-L453 that's where we add shape into shaper.
Similar things happened in another place inside GemmOpBuilder where we called `AddInitializerTransposed` for the second input for a case when bias not provided.

2. Cases involving composed operations where we want to save the shape for later. 
   - Most of the  shape retrieval happened in op_builder_helper.cc file 
   - Here for eltwise operation of BatchNorm op (potentially can get rid of)
   - One place in GemmOpBuilder where we added a SqueezeAll operation. (but for this, it is just a particular log output which if we don't mind can also removed the usage of shaper[].) https://github.com/microsoft/onnxruntime/blob/9edc9465f0258fa4cf83d7975905afa9ce1ec436/onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_builder.cc#L1787-L1794

Everywhere else now are using shape from NodeArg/can call GetShapeInfoFromNodeArg().

but yeah, it does not look ideal now as the way how we get shape now doesn't seem like a unified/general way.